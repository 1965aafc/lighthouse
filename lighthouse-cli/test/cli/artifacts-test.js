/**
 * @license Copyright 2019 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

/* eslint-env jest */

const testServer = require('../fixtures/static-server.js').server;
const run = require('../../run.js');
const getFlags = require('../../cli-flags').getFlags;

/** @type {Array<keyof LH.Artifacts>} */
const ARTIFACTS_TO_ASSERT = [
  'AnchorElements',
  'ImageElements',
  'LinkElements',
  'MetaElements',
  'ScriptElements',
];

/** @type {Array<[(path: string[], value: any) => boolean, (value: any) => any]>} */
const cleaningFunctions = [
  [path => path[0] === 'requestId', () => '<requestId>'],
  [path => path[0] === 'content', value => value && value.slice(0, 100)],
  [(_, value) => /localhost:\d+/.test(value), value => value.replace(/localhost:\d+/, 'localhost')],
  [(_, value) => /^blob:/.test(value), () => '<blob url>'],
];

/**
 *
 * @param {*} value
 * @param {string[]} pathToValue
 */
function cleanArtifactsForSnapshot(value, pathToValue) {
  cleaningFunctions.forEach(([testFn, replaceFn]) => {
    if (testFn(pathToValue, value)) value = replaceFn(value);
  });

  if (Array.isArray(value)) {
    value = value.map((childValue, index) => {
      const childKey = index.toString();
      return cleanArtifactsForSnapshot(childValue, [childKey, ...pathToValue]);
    });
  } else if (typeof value === 'object' && value !== null) {
    Object.entries(value).forEach(([childKey, childValue]) => {
      value[childKey] = cleanArtifactsForSnapshot(childValue, [childKey, ...pathToValue]);
    });
  }

  return value;
}

describe('Artifacts Tests', function() {
  /** @type {LH.RunnerResult} */
  let result;

  beforeAll(async () => {
    await new Promise(resolve => {
      testServer.listen(0, 'localhost', resolve);
    });

    const address = testServer.address();
    const runnerResult = await run.runLighthouse(
      `http://localhost:${address.port}/dobetterweb/dbw_tester.html`,
      getFlags(`--chrome-flags="--headless --no-sandbox" <url>`),
      undefined
    );

    if (!runnerResult) throw new Error('Run failed');
    result = runnerResult;
  }, 30000);

  afterAll(async () => {
    await new Promise(resolve => testServer.close(resolve));
  });

  it('should have created artifacts', () => {
    expect(result).toBeDefined();

    /** @type {Partial<LH.Artifacts>} */
    const artifacts = {};
    ARTIFACTS_TO_ASSERT.forEach(key => (artifacts[key] = result.artifacts[key]));
    expect(cleanArtifactsForSnapshot(artifacts, [])).toMatchSnapshot();
  });
});
