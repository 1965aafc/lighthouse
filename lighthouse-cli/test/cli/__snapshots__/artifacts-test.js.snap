// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Artifacts Tests should have created artifacts 1`] = `
Object {
  "AnchorElements": Array [
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\">",
      "rel": "",
      "target": "_blank",
      "text": "external link",
    },
    Object {
      "href": "",
      "outerHTML": "<a target=\\"_blank\\">",
      "rel": "",
      "target": "_blank",
      "text": "external link",
    },
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\" rel=\\"nofollow\\">",
      "rel": "nofollow",
      "target": "_blank",
      "text": "external link",
    },
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\" rel=\\"noopener nofollow\\">",
      "rel": "noopener nofollow",
      "target": "_blank",
      "text": "external link that uses rel noopener and another unrelated rel attribute",
    },
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\" rel=\\"noreferrer nofollow\\">",
      "rel": "noreferrer nofollow",
      "target": "_blank",
      "text": "external link that uses rel noreferrer and another unrelated rel attribute",
    },
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\" rel=\\"noopener\\">",
      "rel": "noopener",
      "target": "_blank",
      "text": "external link that uses rel noopener",
    },
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\" rel=\\"noreferrer\\">",
      "rel": "noreferrer",
      "target": "_blank",
      "text": "external link that uses rel noreferrer",
    },
    Object {
      "href": "https://www.google.com/",
      "outerHTML": "<a href=\\"https://www.google.com/\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">",
      "rel": "noopener noreferrer",
      "target": "_blank",
      "text": "external link that uses rel noopener and noreferrer",
    },
    Object {
      "href": "http://localhost/dobetterweb/doesnotexist",
      "outerHTML": "<a href=\\"./doesnotexist\\" target=\\"_blank\\">",
      "rel": "",
      "target": "_blank",
      "text": "internal link is ok",
    },
    Object {
      "href": "javascript:void(0)",
      "outerHTML": "<a href=\\"javascript:void(0)\\" target=\\"_blank\\">",
      "rel": "",
      "target": "_blank",
      "text": "",
    },
    Object {
      "href": "mailto:inbox@email.com",
      "outerHTML": "<a href=\\"mailto:inbox@email.com\\" target=\\"_blank\\">",
      "rel": "",
      "target": "_blank",
      "text": "",
    },
  ],
  "ImageElements": Array [
    Object {
      "clientRect": Object {
        "bottom": -261.1875,
        "left": 8,
        "right": 488,
        "top": -318.1875,
      },
      "displayedHeight": 57,
      "displayedWidth": 480,
      "isCss": false,
      "isPicture": false,
      "mimeType": "image/jpeg",
      "naturalHeight": 318,
      "naturalWidth": 480,
      "resourceSize": 24620,
      "src": "http://localhost/dobetterweb/lighthouse-480x318.jpg",
      "usesObjectFit": false,
    },
    Object {
      "clientRect": Object {
        "bottom": 60.8125,
        "left": 8,
        "right": 488,
        "top": -257.1875,
      },
      "displayedHeight": 318,
      "displayedWidth": 480,
      "isCss": false,
      "isPicture": false,
      "mimeType": "image/jpeg",
      "naturalHeight": 318,
      "naturalWidth": 480,
      "resourceSize": 24620,
      "src": "http://localhost/dobetterweb/lighthouse-480x318.jpg",
      "usesObjectFit": false,
    },
    Object {
      "clientRect": Object {
        "bottom": 526.8125,
        "left": 8,
        "right": 819,
        "top": 64.8125,
      },
      "displayedHeight": 462,
      "displayedWidth": 811,
      "isCss": false,
      "isPicture": false,
      "mimeType": "image/gif",
      "naturalHeight": 462,
      "naturalWidth": 811,
      "resourceSize": 934285,
      "src": "http://localhost/dobetterweb/lighthouse-rotating.gif",
      "usesObjectFit": false,
    },
    Object {
      "clientRect": Object {
        "bottom": 655.8125,
        "left": 278,
        "right": 294,
        "top": 639.8125,
      },
      "displayedHeight": 16,
      "displayedWidth": 16,
      "isCss": false,
      "isPicture": false,
      "mimeType": undefined,
      "naturalHeight": 0,
      "naturalWidth": 0,
      "resourceSize": 0,
      "src": "<blob url>",
      "usesObjectFit": false,
    },
  ],
  "LinkElements": Array [
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dobetterweb/dbw_tester.css?delay=100",
      "hrefRaw": "http://localhost/dobetterweb/dobetterweb/dbw_tester.css?delay=100",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/unknown404.css?delay=200",
      "hrefRaw": "http://localhost/dobetterweb/unknown404.css?delay=200",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_tester.css?delay=2200",
      "hrefRaw": "http://localhost/dobetterweb/dbw_tester.css?delay=2200",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_disabled.css?delay=200&isdisabled",
      "hrefRaw": "http://localhost/dobetterweb/dbw_disabled.css?delay=200&isdisabled",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_partial_a.html?delay=200",
      "hrefRaw": "http://localhost/dobetterweb/dbw_partial_a.html?delay=200",
      "hreflang": "",
      "rel": "import",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_partial_b.html?delay=200&isasync",
      "hrefRaw": "http://localhost/dobetterweb/dbw_partial_b.html?delay=200&isasync",
      "hreflang": "",
      "rel": "import",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_tester.css?delay=3000&capped",
      "hrefRaw": "http://localhost/dobetterweb/dbw_tester.css?delay=3000&capped",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "style",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_tester.css?delay=2000&async=true",
      "hrefRaw": "http://localhost/dobetterweb/dbw_tester.css?delay=2000&async=true",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_tester.css?delay=3000&async=true",
      "hrefRaw": "http://localhost/dobetterweb/dbw_tester.css?delay=3000&async=true",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
    Object {
      "as": "",
      "crossOrigin": null,
      "href": "http://localhost/dobetterweb/dbw_tester.css?scriptActivated&delay=200",
      "hrefRaw": "http://localhost/dobetterweb/dbw_tester.css?scriptActivated&delay=200",
      "hreflang": "",
      "rel": "stylesheet",
      "source": "head",
    },
  ],
  "MetaElements": Array [
    Object {
      "content": "",
      "name": "",
    },
    Object {
      "content": "width=device-width, initial-scale=1, minimum-scale=1",
      "name": "viewport",
    },
  ],
  "ScriptElements": Array [
    Object {
      "async": false,
      "content": "
  setTimeout(() => {
    const stylesheet = document.getElementById('amp-style-link');
    styleshe",
      "defer": false,
      "requestId": "<requestId>",
      "source": "head",
      "src": null,
      "type": "text/javascript",
    },
    Object {
      "async": false,
      "content": "/**
 * @license Copyright 2016 Google Inc. All Rights Reserved.
 * Licensed under the Apache License",
      "defer": false,
      "requestId": "<requestId>",
      "source": "head",
      "src": "http://localhost/dobetterweb/dbw_tester.js",
      "type": null,
    },
    Object {
      "async": false,
      "content": null,
      "defer": false,
      "requestId": "<requestId>",
      "source": "head",
      "src": "http://localhost/dobetterweb/empty_module.js?delay=500",
      "type": "module",
    },
    Object {
      "async": false,
      "content": "throw new Error('An error');",
      "defer": false,
      "requestId": "<requestId>",
      "source": "head",
      "src": null,
      "type": "text/javascript",
    },
    Object {
      "async": false,
      "content": null,
      "defer": false,
      "requestId": "<requestId>",
      "source": "head",
      "src": "http://localhost/dobetterweb/fcp-delayer.js?delay=5000",
      "type": null,
    },
    Object {
      "async": false,
      "content": "
  // Ensure gatherers still work when individual elements override '.matches'
  document.getElement",
      "defer": false,
      "requestId": "<requestId>",
      "source": "body",
      "src": null,
      "type": null,
    },
    Object {
      "async": false,
      "content": "window.Error = function(error) { this.stack = 'stacktrace'; };",
      "defer": false,
      "requestId": "<requestId>",
      "source": "body",
      "src": null,
      "type": null,
    },
    Object {
      "async": false,
      "content": "
function stampTemplate(id, location) {
  const template = document.querySelector('#' + id);
  locat",
      "defer": false,
      "requestId": "<requestId>",
      "source": "body",
      "src": null,
      "type": null,
    },
    Object {
      "async": false,
      "content": "/**
* @license
* Copyright Google Inc. All Rights Reserved.
*
* Use of this source code is governed ",
      "defer": false,
      "requestId": "<requestId>",
      "source": "body",
      "src": "http://localhost/zone.js",
      "type": null,
    },
    Object {
      "async": false,
      "content": "/*! jQuery v2.1.1 | (c) 2005, 2014 jQuery Foundation, Inc. | jquery.org/license */
!function(a,b){\\"o",
      "defer": false,
      "requestId": "<requestId>",
      "source": "body",
      "src": "http://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js",
      "type": null,
    },
  ],
}
`;
